pkg/cloudproduct/cloudproduct.go

```diff
@@ -17,38 +17,42 @@ package cloudproduct
 import (
 	"context"
 	"fmt"
+	"sync"
 
 	agonesv1 "agones.dev/agones/pkg/apis/agones/v1"
 	"agones.dev/agones/pkg/client/informers/externalversions"
 	"agones.dev/agones/pkg/cloudproduct/generic"
 	"agones.dev/agones/pkg/cloudproduct/gke"
 	"agones.dev/agones/pkg/portallocator"
 	"agones.dev/agones/pkg/util/runtime"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
 	corev1 "k8s.io/api/core/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/informers"
 	"k8s.io/client-go/kubernetes"
 )
 
-// CloudProduct provides a generic interface that abstracts cloud product
-// specific functionality. Users should call New() to instantiate a
-// specific cloud product interface.
-type CloudProduct interface {
+// The Cloud Product abstraction currently consists of disjoint interfaces:
+// * ControllerHooks() is an interface available to the pkg/controller binaries
+// * api hooks are separated into api group and registered directly with the API
+//
+// Users should bind flags with CloudProductBind{Flags,Env}, then call
+// InitializeCloudProduct with a client to initialize.
+
+// CloudProductControllerHooks provides a generic interface that abstracts cloud product
+// specific functionality for pkg/controller packages.
+type CloudProductControllerHooks interface {
 	// SyncPodPortsToGameServer runs after a Pod has been assigned to a Node and before we sync
 	// Pod host ports to the GameServer status.
 	SyncPodPortsToGameServer(*agonesv1.GameServer, *corev1.Pod) error
 
-	// ValidateGameServer is called by GameServer.Validate to allow for product specific validation.
-	ValidateGameServer(*agonesv1.GameServer) []metav1.StatusCause
-
-	// MutateGameServerPod is called by createGameServerPod to allow for product specific pod mutation.
-	MutateGameServerPod(*agonesv1.GameServer, *corev1.Pod) error
-
 	// NewPortAllocator creates a PortAllocator. See gameservers.NewPortAllocator for parameters.
 	NewPortAllocator(int32, int32, informers.SharedInformerFactory, externalversions.SharedInformerFactory) portallocator.Interface
 }
 
 const (
+	cloudProductFlag = "cloud-product"
+
 	// If --cloud-product=auto, auto-detect
```

Comment:
Comment should start with `AutoDetect`, shouldn't it? I'm trying to work out how this is passing linting?

Reply:
I realize I meant to convert this to unexported anyways, and there was a funny bug before where viper was defaulting to the _function_ `autoDetect` and not the setting. Hurrm.